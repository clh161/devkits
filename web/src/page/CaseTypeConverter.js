// @flow strict
import type { Node } from 'react';
import React, { useState } from 'react';
import { Grid, TextField, Typography } from '@material-ui/core';
import { Helmet } from 'react-helmet';
import lodash from 'lodash';

type CaseTypeKey =
  | 'normal_text'
  | 'camel_case'
  | 'snake_case'
  | 'snake_upper_case'
  | 'kebab_case';

type CaseType = {
  label: string,
  key: CaseTypeKey,
  getTextFromNormalText: (string) => string,
  getNormalText: (string) => string,
};

const CASE_TYPES: Array<CaseType> = [
  {
    label: 'Normal Text',
    key: 'normal_text',
    getTextFromNormalText: (text) => text,
    getNormalText: (text) => text,
  },
  {
    label: 'Camel Case',
    key: 'camel_case',
    getTextFromNormalText: (text) => lodash.camelCase(text),
    getNormalText: (text) => lodash.lowerCase(text),
  },
  {
    label: 'Snake Case',
    key: 'snake_case',
    getTextFromNormalText: (text) => lodash.snakeCase(text),
    getNormalText: (text) => lodash.lowerCase(text),
  },
  {
    label: 'Snake Upper Case',
    key: 'snake_upper_case',
    getTextFromNormalText: (text) => lodash.snakeCase(text).toUpperCase(),
    getNormalText: (text) => lodash.lowerCase(text),
  },
  {
    label: 'Kebab Case',
    key: 'kebab_case',
    getTextFromNormalText: (text) => lodash.kebabCase(text),
    getNormalText: (text) => lodash.lowerCase(text),
  },
];
const DEFAULT_NORMAL_TEXT = 'This is an example text';

const DEFAULT_TEXTS: { [CaseTypeKey]: string } = getTexts(
  'normal_text',
  DEFAULT_NORMAL_TEXT
);

export function getTexts(key: string, text: string): { [CaseTypeKey]: string } {
  const caseType = CASE_TYPES.find((caseType) => caseType.key === key);
  const normalText = caseType?.getNormalText(text) ?? '';
  return CASE_TYPES.reduce((texts, ct) => {
    texts[ct.key] =
      ct.key === key ? text : ct.getTextFromNormalText(normalText);
    return texts;
  }, {});
}

export default function CaseTypeConverter(): Node {
  const [texts, setTexts] = useState<{ [CaseTypeKey]: string }>(DEFAULT_TEXTS);
  return (
    <div>
      <Helmet>
        <title>Case Type Converter</title>
      </Helmet>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <Typography component="h1" variant="h4">
            Case Type Converter
          </Typography>
        </Grid>
        {CASE_TYPES.map((caseType: CaseType) => {
          return (
            <Grid item key={caseType.key} xs={12}>
              <TextField
                fullWidth={true}
                label={caseType.label}
                onChange={(event) => {
                  const { value } = event.target;
                  setTexts(getTexts(caseType.key, value));
                }}
                value={texts[caseType.key]}
              />
            </Grid>
          );
        })}
      </Grid>
    </div>
  );
}
