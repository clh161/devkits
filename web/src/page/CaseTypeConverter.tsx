import { Grid, TextField } from '@material-ui/core';
import React, { ReactElement, useState } from 'react';
import type { CaseType, CaseTypeKey } from './CASE_TYPES';
import { CASE_TYPES } from './CASE_TYPES';
const DEFAULT_NORMAL_TEXT = 'This is an example text';
const DEFAULT_TEXTS: Record<CaseTypeKey, string> = getTexts(
  'normal_text',
  DEFAULT_NORMAL_TEXT
);
export function getTexts(
  key: string,
  text: string
): Record<CaseTypeKey, string> {
  const caseType = CASE_TYPES.find((caseType) => caseType.key === key);
  const normalText = caseType?.getNormalText(text) ?? '';
  // @ts-expect-error
  return CASE_TYPES.reduce((texts, ct) => {
    texts[ct.key] =
      ct.key === key ? text : ct.getTextFromNormalText(normalText);
    return texts;
  }, {});
}
export default function CaseTypeConverter(): ReactElement {
  const [texts, setTexts] =
    useState<Record<CaseTypeKey, string>>(DEFAULT_TEXTS);
  return (
    <Grid container spacing={3}>
      {CASE_TYPES.map((caseType: CaseType) => {
        return (
          <Grid item key={caseType.key} xs={12}>
            <TextField
              fullWidth={true}
              label={caseType.label}
              onChange={(event) => {
                const { value } = event.target;
                setTexts(getTexts(caseType.key, value));
              }}
              value={texts[caseType.key]}
            />
          </Grid>
        );
      })}
    </Grid>
  );
}
