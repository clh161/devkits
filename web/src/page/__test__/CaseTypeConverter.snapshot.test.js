// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import CaseTypeConverter from '../CaseTypeConverter';

it('renders correctly', () => {
  const tree = renderer.create(<CaseTypeConverter />).toJSON();
  expect(tree).toMatchSnapshot();
});
