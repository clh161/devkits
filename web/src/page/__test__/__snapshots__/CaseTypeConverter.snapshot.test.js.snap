// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h1
        className="MuiTypography-root MuiTypography-h4"
      >
        Case Type Converter
      </h1>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Normal Text
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="This is an example text"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Camel Case
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="thisIsAnExampleText"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Snake Case
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="this_is_an_example_text"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Snake Upper Case
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="THIS_IS_AN_EXAMPLE_TEXT"
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Kebab Case
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="this-is-an-example-text"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
